device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/MultiplexingClient.java

```diff
@@ -0,0 +1,252 @@
+     * multiplexed AMQPS_WS connection.
+     * <p>
+     * If the multiplexing client is already open, then this device client will automatically
+     * be opened, too. If the multiplexing client is not open yet, then this device client will not be opened until
+     * {@link MultiplexingClient#open()} is called.
+     * <p>
+     * If the multiplexed connection is already open, then this is an asynchronous operation. You can track the state of your
+     * device session using the {@link DeviceClient#registerConnectionStatusChangeCallback(IotHubConnectionStatusChangeCallback, Object)}.
+     * That callback will execute once your device session has successfully been added to the existing multiplexed connection and
+     * is ready to send telemetry.
+     * <p>
+     * Any proxy settings set to the provided device client will be overwritten by the proxy settings of this multiplexing client.
+     * <p>
+     * The registered device client instance must use the same transport protocol (AMQPS or AMQPS_WS) that this multiplexing client uses.
+     * <p>
+     * The registered device client may have its own retry policy and its own SAS token expiry time, separate from every other multiplexed device.
+     * <p>
+     * The registered device client must use symmetric key based authentication.
+     * <p>
+     * The registered device client must belong to the same IoT Hub as all previously registered device clients.
+     * <p>
+     * If this device client is already registered, then this method will throw a {@link IllegalStateException}
+     * <p>
+     * @param deviceClient The device client to associate with this multiplexing client.
+     */
+    public void registerDeviceClient(DeviceClient deviceClient)
+    {
+        Objects.requireNonNull(deviceClient);
+
+        synchronized (this.lock)
+        {
+            for (DeviceClient currentClient : deviceClientList)
+            {
+                if (currentClient.getConfig().getDeviceId().equalsIgnoreCase(deviceClient.getConfig().getDeviceId()))
+                {
+                    throw new IllegalStateException("Cannot register device client because a device client for that device identity has already been registered");
+                }
+            }
+
+            if (deviceClient.getConfig().getAuthenticationType() != DeviceClientConfig.AuthType.SAS_TOKEN)
+            {
+                throw new UnsupportedOperationException("Can only register a device client if it uses SAS token based authentication");
+            }
+
+            if (deviceClient.getConfig().getProtocol() != this.protocol)
+            {
+                throw new UnsupportedOperationException("Cannot register a device client that uses a different transport protocol than this multiplexing client uses");
+            }
+
+            if (this.protocol == IotHubClientProtocol.AMQPS && this.deviceClientList.size() > 1000)
```

Comment:
How about we make this 1000 a constant, and refer to it in the error message too?

Reply:
Could be a public constant too, to let users know the max number of multiplexed devices

Reply:
I'm a bit hesitant to make it a public variable because the service may change these limits arbitrarily later. We don't want customers using this value directly if we update it from 1,000 to 100,000 to match the service limits. I'd rather just document it in the headers clearly.

Reply:
If the service updates it, and we've been checking it in our code, they'll be broken anyway. We'd just be making the limits we're adhering to public.

I'd say either the client SDK doesn't check these limits, and just lets the service return an error, or we share the limits we're abiding by.

Reply:
Okay, I can make these public then